File: day_23.cpp
#include <cstdint>
#include <cinttypes>
#include <iostream>
#include <exception>
#include <cmath>
#include <optional>

//#include <map>
#include <vector>
#include <algorithm>
#include <tuple>
#include <string>
#include <fstream>

struct Point {
    int64_t x;
    int64_t y;
    int64_t z;
    Point() : x(0), y(0), z(0) {}
    Point(int64_t x_, int64_t y_, int64_t z_) : x(x_), y(y_), z(z_) {}
};

struct Bot : public Point {
    uint64_t r;
    Bot() : Point(), r(0) {}
    Bot(int64_t x_, int64_t y_, int64_t z_, uint64_t r_) : Point(x_, y_, z_), r(r_) {};
};

template<typename T> static inline typename std::make_unsigned<T>::type abs(const T &v) {
    return (v >= T(0)) ? v : -v;
}

template<typename T>
static inline uint64_t manhattan(const T &p) {
    return abs(p.x) + abs(p.y) + abs(p.z);
}

template<typename T1, typename T2>
static inline uint64_t manhattan(const T1 &p1, const T2 &p2) {
    return abs(p1.x - p2.x) + abs(p1.y - p2.y) + abs(p1.z - p2.z);
}

static bool read_input(const std::string &filename, std::vector<Bot> &bots) {
    bots.clear();

    std::ifstream ifs(filename, std::ios::binary);
    if (!ifs.is_open()) {
        fprintf(stderr, "Impossible d'ouvrir le fichier %s\n", filename.c_str());
        return false;
    }

    for (std::string line; std::getline(ifs, line); ) {
        if (line.size() > 0 && line.back() == '\n') {
            line.pop_back();
            if (line.size() > 0 && line.back() == '\r') {
                line.pop_back();
            }
        }
        if (line.size() == 0) {
            continue;
        }

        int64_t x,y,z;
        uint64_t r;
        if (sscanf(line.c_str(), "pos=<%" PRId64 ",%" PRId64 ",%" PRIu64">, r=%" PRId64, &x, &y, &z, &r) != 4) {
            continue;
        }
        bots.emplace_back(x,y,z,r);
    }
    return true;
}

static unsigned int work_p1(const std::string &filename) {
    std::vector<Bot> bots;
    if (!read_input(filename, bots)) {
        fprintf(stderr, "Erreur de lecture de %s\n", filename.c_str());
        return -1;
    }

    // find the bot with the highest range
    uint64_t max_range = 0;
    Bot best_bot;
    for (const auto &p : bots) {
        if (p.r > max_range) {
            max_range = p.r;
            best_bot = p;
        }
    }
    // count all bots in range
    unsigned int in_range = 0;
    for (const auto &p : bots) {
        if (manhattan(p, best_bot) <= max_range) {
            in_range ++;
        }
    }
    return in_range;
}

// apply a scale factor to our space, using closest integer rounding
static void apply_scale(const std::vector<Bot> &in, double factor, std::vector<Bot> &out) {
    out.clear();
    out.reserve(in.size());
    for (const auto &p : in) {
        out.emplace_back(
                std::llround(p.x * factor),
                std::llround(p.y * factor),
                std::llround(p.z * factor),
                std::llround(p.r * factor));
    }
}

static uint64_t work_p2(const std::string &filename) {
    std::vector<Bot> bots;
    if (!read_input(filename, bots)) {
        fprintf(stderr, "Erreur de lecture de %s\n", filename.c_str());
        return -1;
    }

    // find our boundaries and ranges on each axis
    uint64_t range_max;
    {
        int64_t min_x, max_x, min_y, max_y, min_z, max_z;
        auto it = bots.begin();
        min_x = max_x = (*it).x;
        min_y = max_y = (*it).y;
        min_z = max_z = (*it).z;
        for (; it != bots.end(); ++it) {
            const auto &p = *it;
            min_x = std::min(min_x, p.x); max_x = std::max(max_x, p.x);
            min_y = std::min(min_y, p.y); max_y = std::max(max_y, p.y);
            min_z = std::min(min_z, p.z); max_z = std::max(max_z, p.z);
        }
        range_max = std::max(max_x - min_x, std::max(max_y - min_y, max_z - min_z));
    }

    std::optional<unsigned int> best_in_range; // current max bots in range
    uint64_t best_distance = uint64_t(-1);  // current best distance from (0,0,0)
    Point best_point;                       // current best point
    uint64_t scan_width = 300;              // for the first pass, we scale the space to a 300x300x300 space
    Point current_pos;                      // current center point (previous best)
    

    auto find_best_point = [&](const std::vector<Bot> &bots) {
        Point p;
        // search a scale_width*scale_width*scale_with cube centered on current_pos
        for (p.x = current_pos.x - scan_width / 2; p.x <= int64_t(current_pos.x + scan_width / 2); ++p.x) {
            for (p.y = current_pos.y - scan_width / 2; p.y <= int64_t(current_pos.y + scan_width / 2); ++p.y) {
                for (p.z = current_pos.z - scan_width / 2; p.z <= int64_t(current_pos.z + scan_width / 2); ++p.z) {
                    unsigned int in_range = 0;
                    for (const auto &b : bots) {
                        if (manhattan(b, p) <= b.r) {
                            in_range++;
                        }
                    }
                    uint64_t p_dist = manhattan(p);
                    if (!best_in_range || (in_range > best_in_range) || (in_range == best_in_range && p_dist < best_distance)) {
                        best_distance = p_dist;
                        best_in_range = in_range;
                        best_point = p;
                    }
                }
            }
        }
    };

    double scale_factor = double(scan_width) / range_max;
    decltype(bots) scaled_bots;

    // the scale factor is double each pass
    while (scale_factor < 1) {
        apply_scale(bots, scale_factor, scaled_bots);
        find_best_point(scaled_bots);

        printf("<%" PRId64 ",%" PRId64 ",%" PRId64 "> %" PRIu64 " : %u\n", best_point.x, best_point.y, best_point.z, best_distance, *best_in_range);
        fflush(stdout);

        if (scale_factor * 2 < 1) {
            current_pos = Point{best_point.x * 2, best_point.y * 2, best_point.z * 2};
            best_in_range.reset();
            scale_factor *= 2;

            // for the following passes, reduce the sampling size
            scan_width = 75;
        }
        else {
            current_pos = Point{int64_t(best_point.x / scale_factor), int64_t(best_point.y / scale_factor), int64_t(best_point.z / scale_factor)};
            break;
        }
    }

    // final pass without scaling
    scan_width = std::min(scan_width, range_max); // restrict for the smol test case
    find_best_point(bots);

    printf("<%" PRId64 ",%" PRId64 ",%" PRId64 "> %" PRIu64 " : %u\n", best_point.x, best_point.y, best_point.z, best_distance, *best_in_range);
    fflush(stdout);

    return best_distance;
}

static bool test_p1() {
    return work_p1("input_23_test_p1") == 7;
}

static bool test_p2() {
    return work_p2("input_23_test_p2") == 36;
}

int main() {
    if (!test_p1()) {
        fprintf(stderr, "Erreur test_p1\n");
    }
    printf("%u\n", work_p1("input_23"));

    if (!test_p2()) {
        fprintf(stderr, "Erreur test_p2\n");
    }
    printf("%" PRIu64 "\n", work_p2("input_23"));

    return 0;
}

File: day_23.py
#!/usr/bin/python3
#encoding: UTF-8

# this one does the job, but it's not satisfying as it's a randomized search

import fileinput
import sys
import re
import itertools
import random
random.seed()

if len(sys.argv) == 1:
    sys.argv += ["input_23"]

sample_input_p1 ="""pos=<0,0,0>, r=4
pos=<1,0,0>, r=1
pos=<4,0,0>, r=3
pos=<0,2,0>, r=1
pos=<0,5,0>, r=3
pos=<0,0,3>, r=1
pos=<1,1,1>, r=1
pos=<1,1,2>, r=1
pos=<1,3,1>, r=1""".splitlines()

sample_input_p2 = """pos=<10,12,12>, r=2
pos=<12,14,12>, r=2
pos=<16,12,12>, r=4
pos=<14,14,14>, r=6
pos=<50,50,50>, r=200
pos=<10,10,10>, r=5""".splitlines()

def read_input(lines):
    bots = {}
    re_int = re.compile("[-]?[0-9]+")
    for line in lines:
        line = line.strip()
        if len(line) == 0:
            continue
        x,y,z,r = list(map(int, re_int.findall(line)))
        bots[x,y,z] = r
    return bots

def manhattan(p1, p2):
    return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1]) + abs(p1[2] - p2[2])

def work_p1(bots):
    max_range = max(bots.values())
    best_bot = next(b for b, r in bots.items() if r == max_range)
    in_range = 0
    for b, r in bots.items():
        if manhattan(best_bot, b) <= max_range:
            in_range += 1
    return in_range

def test_p1():
    bots = read_input(sample_input_p1)
    assert work_p1(bots) == 7
test_p1()

def p1():
    bots = read_input(fileinput.input())
    print(work_p1(bots))
p1()


def cube(factor, w=3):
    if w==3:
        return itertools.product((- factor, 0, factor), repeat=3)
    elif (w&1) == 1:
        # we can do a denser grid
        grid = itertools.product(range(-(w//2)*factor, (w//2+1)*factor, factor), repeat=3)
        return [tuple(v//w for v in e) for e in grid]
    else:
        print("w pair")
        sys.exit(25)

def rand_points(factor, w=3, count=27):
    s = set(cube(factor, w))
    for c in range(0, count):
        x = random.randint(-factor, factor)
        y = random.randint(-factor, factor)
        z = random.randint(-factor, factor)
        s.add((x,y,z))
    return s

def work_p2(bots):
    # we will search using a sampling grid, reducing its size after each iteration
    # for each point of the grid, we count how many bots have the point in range
    # we may need multiple runs to find the solution.
    # the "grid" is a set of random points and regular points on a cube
    # I guess it does work because the ranges are huge but I'm not satisfied as
    # it will often not converge towards the best point (although the distance is
    # usually the best)
    
    # let's start by finding the max size
    min_x, max_x = min(x for x,y,z in bots), max(x for x,y,z in bots)
    min_y, max_y = min(y for x,y,z in bots), max(y for x,y,z in bots)
    min_z, max_z = min(z for x,y,z in bots), max(z for x,y,z in bots)

    window_size = max(abs(min_x), abs(max_x), abs(min_y), abs(max_y), abs(min_z), abs(max_z))
    reduce_factor = 0.8
    
    # count the number of nanobots which have p in their range
    def have_in_range(p):
        nonlocal bots
        in_range = 0
        for b,r in bots.items():
            if manhattan(p, b) <= r:
                in_range += 1
        return in_range
    
    grid_size=5
    
    overall_best = None
    
    for iteration in range(5):
        best_in_range = -1      # best number of nanobots in range
        best_distance = None    # best distance from (0,0,0)
        window = window_size
        current_pos = (0,0,0)
        for it in range(3):
            while True:
                for position in rand_points(window, grid_size, grid_size**3):
                    position = tuple(current_pos[i] + position[i] for i in (0,1,2))
                    in_range = have_in_range(position)
                    this_dist = manhattan(position, (0,0,0))
                    if (in_range > best_in_range) or (in_range == best_in_range and this_dist < best_distance):
                        best_in_range = in_range
                        best_distance = this_dist
                        best_position = position
                        #print(window, position, in_range, this_dist)
                current_pos = best_position
                if window == 1:
                    break
                n_window = int(reduce_factor * window)
                if n_window == window:
                    n_window = window - 1
                if n_window == 0:
                    n_window = 1
                window = n_window
        print(current_pos, best_in_range, best_distance)
        overall_best = best_distance if overall_best == None else min(overall_best, best_distance)
    return overall_best

def test_p2():
    bots = read_input(sample_input_p2)
    assert work_p2(bots) == 36
test_p2()

def p2():
    bots = read_input(fileinput.input())
    print(work_p2(bots))
p2()
